### PR TITLE
Static RPCs & Authority RPCs

### PR DESCRIPTION
**Pull request by kurozael**
This might not be how you wanna handle it but just whipped this up for the discussion really.

## Authority Attribute
Similar to Broadcast but it'll only run the RPC on the owner (if it has an owner, or the host if it doesn't).

### Requirements
Will require a way to send messages directly to a connection id for example `GameNetworkSystem.Send( Guid connectionId, T obj )`. When the method is called and you are the proxy, call Broadcast to send to the host first and then when the host gets it they'll send directly to the owner (if they are also a proxy)

## Static RPCs
Can add `[Broadcast]` to any static method and have it able to be called by anyone in the same way that it works for GameObjects and Components.

All available static methods that have this attribute are cached on load and hotload to create a dictionary lookup of indexes (`int`) to full method name (`string`). When sending messages the `int` is used instead of the `string`.

## Other Stuff
When writing stuff to index static methods with an int for sending over the network, I applied this to regular `[Broadcast]` methods too so sending a `string` for those is no longer needed either.

## Future
I noticed you have a TODO comment for handling multiple components of the same type when calling RPCs. Did you have any ideas on how you want to identify them? We could generate a GUID for each component and send that when calling the RPC although that would be 16 bytes (does it matter?)

### CODE REVIEW

**Comment by garrynewman** [comment]
This stuff would be a lot easier to digest if you PR'd one thing at a time.

Static RPCs - yep fine, works for me.
Ints for methods - yep fine. I left them as strings for debugging during this phase but yep.

The Authority is a bit more confused. Like you say, we want to message the host, and the network model is ambiguous. What I think we should be doing is this:

* Call `SceneNetworkSystem.Instance.Send( Network.OwnerId, msg );`
* If capable (because we're p2p, or host) sends directly to that connection.
* If not then sends to the host, who then sends it to the connection.

The distinction between what you're doing and this, is that the implementation should be deciding how to get that message to that client, not the RPC code.

Some other notes..

* TryFindMethodIndex is always going to throw, so just call it FindMethodIndex and have that method throw instead of checking in each RPC method.
* Lets add those attributes to WrapStaticMethod so it doesn't show up in intellisense etc
* I don't really like that we're building and trying to maintain a set of dictionaries for the network methods etc. Would much rather it looks methods up on a cache miss so it builds naturally over time instead of all at once, every hotload. Hotload would just clear the cache.

**Comment by kurozael** [comment]
That's great thanks for the info and sorry about the single PR - I started with the static RPCs and kind of snowballed from there and things started tying together.

**Comment by dictateurfou** [comment]
hey I would like to make a suggestion if it is possible.
On the old system the principle was quite similar we had to put static methods it's cool to be able to call them directly but that doesn't offer only advantages so I would like to share my experience.

What could be good is to have an event system with names

This would offer several advantages already the possibility for plugins to create events which could be recorded on any function and thus execute several functions in multiple places without having to use a static function.

Of course we can very well create this event system around the RPCs but if it were integrated into the game it would be really cool

for example if a player throws an item and someone wants to add an event to execute an action each time an item is thrown they could simply record an event

exemple with code
```csharp
public partial class PlayerManager
{
    [RegisterServerEvent("player:trashItem")]
    public void TrashItem( IClient client, string stackId, int qty )
    {
			var player = (client.Pawn as Player);
			var container = TrashContainerManager.CreateContainer( player.Position + player.Rotation.Forward * 40 );
			player.PlayerData.Inventory.Transfert(stackId,qty,container.Inventory);
    }
}

public partial class AnotherPlayerManager
{    
    [RegisterServerEvent("player:trashItem")]
     public void TrashItem( IClient client, string stackId, int qty )
    {
        //make another action
    }
}

//And for call the function in client to server you just want to write
EventCommand.TriggerServerEvent("player:trashItem", stackId, Quantity)
//for Server To Client
EventCommand.TriggerClientEvent( client, "ui:open", "CharCreator", true );
```

this is just a suggestion it is entirely possible to build your own system for it but it would add a plus if it was directly integrated into the game after that it can be an additional addition and not a replacement of the rpc

**Comment by kurozael** [comment]
Alright cool @garrynewman I think I've sorted all concerns though it relies on facepunch/sbox#1363 for the targeted messages. I'd wanna get that one sorted out and merged so I can re-test everything again on a real build.

File: code/GameEngine/Networking/Rpc.cs [hunk 1]
```diff
@@ -90,4 +143,94 @@ private static void InvokeRpc( in ObjectMessageMsg message, in TypeDescription t
...
+		var instanceMethods = TypeLibrary.GetTypes()
+			.SelectMany( x => x.Members )
+			.OfType<MethodDescription>()
+			.Where( x => !x.IsStatic && ( x.Attributes.OfType<BroadcastAttribute>().Any() || x.Attributes.OfType<AuthorityAttribute>().Any() ) )
+			.Select( ( m ) => m.Name );
+
+		return staticMethods
+			.Concat( instanceMethods )
+			.Distinct()
+			.Order();
```
Comment (garrynewman) [inline code/GameEngine/Networking/Rpc.cs:234]: This method shouldn't exist. We should be able to get an Identifier from MethodDescription, and look up straight from TypeLibrary. We shouldn't need to duplicate this stuff when it's already all stored in TypeLibrary.

Comment (kurozael) [reply code/GameEngine/Networking/Rpc.cs:234]: Neither `MemberDescription` or `MethodDescription` have an identifier - or are ya saying we should add it in?

Comment (kurozael) [reply code/GameEngine/Networking/Rpc.cs]: I think my latest commit should be good then - got rid of that in favor of using a `FastHash()` of the fully qualified method name 👍

Comment (garrynewman) [reply code/GameEngine/Networking/Rpc.cs]: TypeLibrary already has this built in via MemberDescription.Ident - we shouldn't be re-inventing the wheel here.

**Comment by kurozael** [comment]
Should be kind of chilling now with the changes we discussed @garrynewman - needs a look over to make sure we're happy with it. Heavily relies on https://github.com/Facepunch/sbox/pull/1371

**Pull request closed without merging**